Create Fedor Korotkov podcast post (#5200)
sourcegraph/about: podcast/2.19.md

Just at some point, I can be sitting at the park, and then I'm just like, "Oof," and then I'm freezing. And then I'm like, "Oh, oh. Yeah, yeah." And then… so I'm analyzing. Yeah, that's a little bit weird.

**Beyang Liu:**

That is so true.

**Fedor Korotkov:**

Yeah. But then you end up with great ideas. And hopefully it's somewhat popular. I'm always very modest about saying Cirrus is popular, Cirrus is the best and so on. But it's somewhat popular, so people really enjoy it. There is really good retention. So people, when they start using it, they don't stop using it. This is the best compliment that you can do, to just continue using it.

**Beyang Liu:**

That's awesome. So I want to talk a bit more about the—you touched upon this a bit—about the different approach that Cirrus takes to CI, and specifically the design decisions and the architecture. Can you talk a bit about, from your point of view, what has been the evolution of CI systems? And what is the "old way" or the conventional way of doing things? And what is the way that Cirrus is approaching it that's different?

**Fedor Korotkov:**

> Yeah, sure. I have a really good blog post about it. I think maybe we can include it in the show notes to the podcast.

Reviewer: We should make sure the blog post mentioned is in the show notes

**Beyang Liu:**

What about startup times, and also caching? Because I could see the argument being made: if you have a long-running worker, it can start up faster than a cold start VM. And then maybe there's a local cache that it keeps around that could help make things more efficient too.

**Fedor Korotkov:**

Yeah. That's for sure. There is a trade-off for virtual machines. So if you want to run a virtual machine, then there is probably 20 to 40 seconds for the spin-up time, which is mostly acceptable if you need a VM.

For containers, the situation is different. Containers you can use Kubernetes, and Kubernetes, in our testing, can spin up a new container in 2 seconds, if all the things are aligned together. But even if not, I constantly measure P99 for the startup time of the containers and on Cirrus it's 7 seconds. So 99 tasks of 100 will start in less than 7 seconds.

**Beyang Liu:**

Wow.

**Fedor Korotkov:**

> And there is no work from our side. We just utilize Kubernetes. And, in our case, we utilize GKE— Google's Kubernetes, but Cirrus also supports AWS, CKS, Amazon Kubernetes, and even Oracle Cloud Kubernetes. So you can connect your own infrastructure and run your CI, only your own cluster within your own infrastructure, which is also super beneficial. You can get all the discounts you have from the cloud provider. You have more control over configurations, like some like edge cases and so on.

Reviewer: ```suggestion
And there is no work from our side. We just utilize Kubernetes. And, in our case, we utilize GKE— Google's Kubernetes, but Cirrus also supports AWS, CKS, Amazon Kubernetes, and even Oracle Cloud Kubernetes. So you can connect your own infrastructure and run your CI, on your own cluster within your own infrastructure, which is also super beneficial. You can get all the discounts you have from the cloud provider. You have more control over configurations, like some like edge cases and so on.
```

**Fedor Korotkov:**

I don't think … of course GitHub Actions right now is the major CI on the GitHub platform. That's for sure. There were actually two versions of the GitHub Actions. There were GitHub Actions before the Microsoft acquisition and GitHub Actions after. So I was very worried about the GitHub Actions prior to the Microsoft acquisition—back then there was Actions better, that I was part of. But then the acquisition happened and basically GitHub Actions right now is just a fork of Azure pipelines.

**Beyang Liu:**

Interesting.

**Fedor Korotkov:**

Yeah. So ...

**Beyang Liu:**

So we kept the brand, but it's completely different from the old Actions?

**Fedor Korotkov:**

> In the very beginning, if you name the dash, you will see this… it's like “Azure blah, blah, blah, kernel.” So it's like the Azure pipeline.

Reviewer: I'm not familiar with the phrase "name the dash" - are we sure of that one?

But first of all, they are one size. I think they're introducing more sizes. So more than two CPUs, then there is a niche. But my main problem with the—it's not only GitHub Actions, but these kind of CIs—is that they have a base virtual machine that people use that can change. And it's the 21st century: we're always talking about infrastructure as code, reproducible infrastructure and so on and so forth. And I think many people were in a situation that they had some side project, or some project at work that they have, that they haven't touched in a couple of months. They come, they do some pull request, and it just fails. And you just don't know why. It just randomly fails because some third-party package got updated.

**Beyang Liu:**

> In theory, it shouldn't fail because it's all containerized and blah, blah, blah. But if you're changing out the underlying thing, at the end of the day, it's the abstraction battery's not perfect. Right.

Reviewer: I think this is "abstraction boundary" but the second word is a bit unclear.

They wanted to first test Podman on specific versions of operating systems, with a specific kernel version. Because it's highly integrated and they just enjoyed it. And I'm really glad that they provided a lot of good feedback about their use case for running CI.

Who else? There is SonarSource. SonarSource is a … you should be probably familiar; it's …

**Beyang Liu:**

Yeah, yeah, super familiar. So these are all great companies and logos. It's pretty incredible that you have such a diverse set of teams using Cirrus. Can you talk a little bit about what's coming up in the Cirrus roadmap? Anything exciting that you're able to share?

**Fedor Korotkov:**

Yeah. I don't want to jinx it too much, but there is definitely coming … we are pretty late to the game, but we are launching capability to basically SSH into the running instances. Because Cirrus supports that many environments, so you can have Azure, AWS, Google Cloud, or Oracle Cloud. You can have your own hardware under firewall. We had to be a little bit tricky about this terminal support.

> And, yeah, it's coming in the next few days. There will be a blog post we can probably also include. And we implemented it as an open-source Go package. So basically this package will allow anyone to integrate SSH access to their Go application. So, yeah, we can include it.

Reviewer: Check for this blog post.

And one other thing, I think after Apple released their Apple Silicon …

**Beyang Liu:**

M1?

**Fedor Korotkov:**

Nowadays, everyone’s like, "We need to have ARM. We need to have ARM." And there were a lot of requests about ARM instances. And there is definitely something coming very soon.

**Beyang Liu:**

Cool.

**Fedor Korotkov:**

So I need to mention that Cirrus already supports ARM. So you can run on your own hardware or your own cloud. So you can have Oracle Cloud running ARM or AWS running ARM.

And we are very soon introducing managed instances. It will be, of course, free for open source. And open source will be able to make sure that their projects are very optimized for ARM processors.

**Beyang Liu:**

> Cool. Would it be safe to say that the two biggest benefits of Cirrus is … there's kind of two, and they're separate, at least in my mind. One is just developer experience, the fact that all this is configured by hitting cloud APIs. And you don't have to manage a bunch of infrastructure, in order to stand up CI and scale it as your team grows. So developer experience, one.

Reviewer: ```suggestion
Cool. Would it be safe to say that the two biggest benefits of Cirrus are … there's kind of two, and they're separate, at least in my mind. One is just developer experience, the fact that all this is configured by hitting cloud APIs. And you don't have to manage a bunch of infrastructure, in order to stand up CI and scale it as your team grows. So developer experience, one.
```

So even if people will decide to move from Cirrus, there is no need to convert existing Cirrus configuration into some other CI system, because Cirrus CLI can run locally.

**Beyang Liu:**

Oh, wow. So I can use this to run my build locally?

**Fedor Korotkov:**

Yes.

**Beyang Liu:**

I feel like that's useful even if I'm not concerned about the switching thing. Because there's so many times where the thing failed in the build environment, but when I run the test locally, it seems to succeed. Now you're saying you could use Cirrus CLI to reproduce the CI environment on your local machine.

**Fedor Korotkov:**

Yeah. Only Docker images, and this …

**Beyang Liu:**

Yeah, yeah, of course.

**Fedor Korotkov:**

… and the most popular instances that Cirrus runs. That's why we were very late with SSH access, because you always could use Cirrus CLI to run your CI locally to reproduce things. And 99.9% of the cases, you will be able to reproduce it.

> But then there is, of course, problems when you don't have, for example, FreeBSD access, or you don't have macro AST. And your tests are only failing on one particular virtual machine, for example. Then it's more convenient to just rerun things in the browser you have. With the terminal access in the browser, you can figure things out and then fix it.

Reviewer: ```suggestion
But then there is, of course, problems when you don't have, for example, FreeBSD access, or you don't have macOS. And your tests are only failing on one particular virtual machine, for example. Then it's more convenient to just rerun things in the browser you have. With the terminal access in the browser, you can figure things out and then fix it.
```

---

> title: 'Creating one CI to rule them all, with Fedor Korotkov, founder and CTO of Cirrus Labs'

Reviewer 2: @nickmyyz @csbailey5t Cirrus Labs should be one word: CirrusLabs

publishDate: 2022-03-15T10:00-07:00
tags: [podcast]
slug: fedor-korotkov
published: true
---

<!-- START AUDIO -->

<audio className="object-center" src="https://www.buzzsprout.com/1097978/10038427-building-the-code-editor-dreams-are-made-of-with-max-brunsfeld-co-founder-of-zed.mp3?download=true" controls preload="none"></audio>

<!-- END AUDIO -->

<!-- START GUESTS -->
<span>
Fedor Korotkov, Beyang Liu
</span>
<!-- END GUESTS -->

<!-- START SUMMARY -->

> Why can’t one CI scale alongside a company–from startup to enterprise? In this episode, Fedor Korotkov, founder and CTO of Cirrus Labs, joins Beyang Liu, co-founder and CTO of Sourcegraph, to talk about how, as a student back in 2009, he developed a photo app that earned him almost $2,000 a month, share the time he applied to be an intern at Twitter but ended up with a full-time job, and explain how six months of “funemployment” led to the building and founding of Cirrus CI–the one CI to rule them all. Along the way, Fedor explains how Cirrus CI, with Kubernetes, can spin up a new container in two seconds.

Reviewer 2: @nickmyyz  @csbailey5t same comment here. The name should be "CirrusLabs," one word. I think the other mentions are all just "Cirrus" in the transcript, so the rest should be okay.

**Beyang Liu:**

If they want to try it out, maybe install it on one of their open-source repositories, how do they go about doing that?

**Fedor Korotkov:**

They can just Google Cirrus CI, and it's as easy as installing a Cirrus CI application from GitHub marketplace. And we have a very nice integration with GitHub. We support annotation. So all the failing things will be highlighted right on the PR.

We support fancy GitHub Checks API, where you … in most cases, you don't even need to go to Cirrus UI, because everything is available on your pull request, or UI inside GitHub. So, yeah, it's super easy. And then we have a lot of examples of how to use, and we support, as we discussed, a lot of use cases and scenarios to test things.

**Beyang Liu:**

Awesome. Well, thanks so much, Fedor, for taking the time to be on the show. Real pleasure chatting with you.

**Fedor Korotkov:**

Thank you, Beyang.

> 

Reviewer 2: @nickmyyz @csbailey5t I noticed we did not include the "This transcript has been lightly edited for clarity and readability." sentence at the end. Can you please add this?
